palette/src/random_sampling/cone.rs

```diff
@@ -0,0 +1,173 @@
+use core::marker::PhantomData;
+
+use crate::float::Float;
+use crate::rgb::RgbSpace;
+use crate::{from_f64, FloatComponent, Hsl, Hsv};
+
+// Based on https://stackoverflow.com/q/4778147 and https://math.stackexchange.com/q/18686,
+// picking A = (0, 0), B = (0, 1), C = (1, 1) gives us:
+//
+// (  sqrt(r1) * r2  ,  sqrt(r1) * (1 - r2) + sqrt(r1) * r2  ) =
+// (  sqrt(r1) * r2  ,  sqrt(r1) - sqrt(r1) * r2 + sqrt(r1) * r2  ) =
+// (  sqrt(r1) * r2  ,  sqrt(r1)  )
+//
+// `sqrt(r1)` gives us the scale of the triangle, `r2` the radius.
+// Substituting, we get `x = scale * radius, y = scale` and thus for cone
+// sampling: `scale = powf(r1, 1.0/3.0)` and `radius = sqrt(r2)`.
+pub fn sample_cone<T: FloatComponent>(r1: T, r2: T) -> (T, T) {
+    let h = Float::cbrt(r1);
+    let r = h * Float::sqrt(r2);
+
+    (h, r)
+}
+
+pub fn invert_cone_sample<S, T>(color: Hsv<S, T>) -> (T, T)
+where
+    T: FloatComponent,
+    S: RgbSpace,
+{
+    let sqrt_r2 = if color.value > from_f64::<T>(0.0) {
+        color.saturation / color.value
+    } else {
+        from_f64::<T>(0.0)
+    };
```

Comment:
Isn't `sqrt_r2` just the saturation? HSV is kind of just the "lower cone" scenario for HSL, but without the scaling. So when it's sampled, `h` gets cancelled out by normalization of the saturation to [0, 1]. That makes this whole statement disappear, although it's probably a good idea to clamp saturation to not go below 0 in both HSV and HSL. Otherwise there will be some weird results for negative numbers.